### ISSUE TITLE
[BigQuery]  Couldn't find closing bracket for opening bracket.

### ISSUE THREAD

**Issue by barrachri** [sqlfluff/sqlfluff#856]
Hi there!

I am running sqlfluff on a BigQuery update query and the linter throws a `Couldn't find closing bracket for opening bracket.
` error.

## lint 

```sql
CREATE TEMPORARY TABLE final_rows AS
(
    WITH source_table AS
    (
        SELECT
        sd.*,
        ccdv.version_id
        FROM
        source_data AS sd
        JOIN current_carrier AS ccdv
        ON ccdv.origin_country_id=sd.origin_country_id
        WHERE
        ccdv.destination_country_id=sd.destination_country_id
        AND ccdv.carrier_id=sd.carrier_id
        AND sd.order_date < sd.ship_date
        AND sd.ship_date < sd.delivery_date
    )
    SELECT
    sdp.po_number,
    cspo.station AS origin_station,
    cspd.station AS destination_station
    FROM
    source_table AS sdp
    LEFT JOIN `another_table` AS cspo
    ON sdp.origin=cspo.id
    LEFT JOIN `yet_another_table` AS cspd
    ON sdp.destination_country_id=cspd.id
    WHERE
    sdp.carrier_id=cspo.id
    AND sdp.carrier=cspo.version
    AND sdp.carrier_id=cspd.id
)
;
```

## Expected Behaviour

No errors

## Observed Behaviour
```
L:   4 | P:   5 |  PRS | Couldn't find closing bracket for opening bracket.

```

## Steps to Reproduce

```
sqlfluff lint file_name
```

## Version
Include the output of `sqlfluff --version` along with your Python version

## Configuration
```
[sqlfluff]
verbose = 0
dialect = bigquery
templater = python
```

**Comment by barrywhart** [comment]
I think this is the same as #835. The BigQuery dialect has issues parsing queries with `<` or `>` inside parentheses.

I have a PR, #836, which addresses that issue.

**Comment by barrywhart** [comment]
I tested and reproduced this on the latest `master` as of 2021-03-14. I believe the issue is that SQLFluff's grammar for `CREATE TABLE AS` is not supporting CTEs. If I remove the CTE, it parses okay. I pasted the test query into the BigQuery web editor, and it parses okay, so I agree this is a bug.

I'll see if I can narrow down the issue with a smaller query or something specific about the grammar.

**Comment by barrywhart** [comment]
Updates:
It also fails with (apparently) the same error if `dialect = ansi`

This simplified version of the query also fails to parse:
```
CREATE TABLE final_rows AS
(
    WITH source_table AS
    (
        SELECT
        *
        FROM
        source_data AS sd
    )
    SELECT
    sdp.po_number
    FROM
    source_table AS sdp
)
```

Moving the CTE to a nested query works:
```
CREATE TABLE final_rows AS
(
    SELECT
    sdp.po_number
    FROM
    (SELECT * FROM source_data) AS sdp
)
```

The ANSI grammar _appears_ to support this syntax:
* `CREATE TABLE AS` can be [followed by a `SelectableGrammar`](https://github.com/sqlfluff/sqlfluff/blob/master/src/sqlfluff/core/dialects/dialect_ansi.py#L1832)
* `SelectableGrammar` [supports `WithCompoundStatementSegment`](https://github.com/sqlfluff/sqlfluff/blob/master/src/sqlfluff/core/dialects/dialect_ansi.py#L1560L1562)

So the problem seems a bit subtle.

**Comment by barrywhart** [comment]
I found the problem! I noticed that if I remove the parentheses around the query, it parses:
```
CREATE TABLE final_rows AS

    WITH source_table AS
    (
        SELECT
        *
        FROM
        source_data AS sd
    )
    SELECT
    source_table.po_number
    FROM
    source_table
```

The issue is that `CREATE TABLE AS SELECT` needs to allow parentheses. A basic `SELECT` (no CTE) works because (I think) you're allowed to surround a query with parentheses, like with nested queries.

Question: With `CREATE TABLE AS SELECT`, are parentheses *required* or *optional*?

**Comment by barrywhart** [comment]
Answer: `CREATE TABLE AS SELECT` should _not_ require parentheses, but they are permitted.

https://cloud.google.com/bigquery/docs/reference/standard-sql/data-definition-language#query_statement

**Comment by barrywhart** [comment]
Note that this is related to #650: "Are we optionally bracketing where we should be, and should we add an OptionallyBracketed segment?"

**Comment by barrachri** [comment]
Thanks @barrywhart, I'll modify the query 👍

**Comment by barrywhart** [comment]
@barrachri: I'm sorry -- that message about parentheses was for me (the person working on a fix), not for you. But yes, removing the parentheses should serve as a workaround until the fix is merged and released (PR #864).
